src/main/java/com/sequenceiq/cloudbreak/service/stack/handler/StackUpdateSuccessHandler.java

```diff
@@ -59,22 +66,46 @@ public void accept(Event<StackUpdateSuccess> t) {
 
     }
 
-    private void terminateMetaDataInstances(StackUpdateSuccess updateSuccess, Long stackId, Set<String> instanceIds, InstanceGroup instanceGroup) {
-        int nodeCount = instanceGroup.getNodeCount() - instanceIds.size();
-        stackUpdater.updateNodeCount(stackId, nodeCount, updateSuccess.getHostGroup());
-
+    private List<InstanceMetaData> getMetaData(InstanceGroup instanceGroup, Set<String> instanceIds) {
+        List<InstanceMetaData> result = new ArrayList<>(instanceIds.size());
         for (InstanceMetaData metadataEntry : instanceGroup.getInstanceMetaData()) {
             if (instanceIds.contains(metadataEntry.getInstanceId())) {
-                long timeInMillis = Calendar.getInstance().getTimeInMillis();
-                metadataEntry.setTerminationDate(timeInMillis);
-                metadataEntry.setTerminated(true);
+                result.add(metadataEntry);
             }
         }
+        return result;
+    }
+
+    private void terminateMetaDataInstances(StackUpdateSuccess updateSuccess, long stackId,
+            List<InstanceMetaData> instancesMetaData, InstanceGroup instanceGroup) {
+        int nodeCount = instanceGroup.getNodeCount() - instancesMetaData.size();
+        stackUpdater.updateNodeCount(stackId, nodeCount, updateSuccess.getHostGroup());
+
+        for (InstanceMetaData instanceMetaData : instancesMetaData) {
```

Comment:
To eliminate this loop I think that could be a proper solution if we terminate metadata-s when we "collect them in 'getMetaData'" function and the returned list of instance metadata could be used by 'removeNodesFromConsul' function.